### PR TITLE
List contacts gui

### PR DESCRIPTION
**Pull request by cnak**

### CODE REVIEW

File: WPFGui/GuiContactManagement/AddContact/AddContactInputView.cs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+                MinWidth = 80,
+                MaxHeight = 30,
+                MaxWidth = 90
+            };
+
+            okButton.Click += btnDialogOk_click;
+            return okButton;
+        }
+
+        private void btnDialogOk_click(object sender, RoutedEventArgs e)
```
Comment (c-j-j) [inline WPFGui/GuiContactManagement/AddContact/AddContactInputView.cs]: does this method have to be named in such a technical way?

File: ContactModel/ContactModel/ContactValidator.cs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        }
+
+        private static bool IsNotValidLength(string name)
+        {
+            return name.Length < 2;
+        }
+
+        private static bool IsValidName(string name)
+        {
+            if (ContainsNumber(name) || IsNotValidLength(name))
```
Comment (c-j-j) [inline ContactModel/ContactModel/ContactValidator.cs]: there's no need for an if statement here.. just return the result of a boolean expression.

File: ContactModel/ContactModel/ContactValidator.cs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        {
+            return postcode.Length > 1;
+        }
+
+        private static bool IsValidAddressLength(string address)
+        {
+            return address.Length > 3;
+        }
+
+        private static bool ContainsNumber(string contactInfo)
```
Comment (c-j-j) [inline ContactModel/ContactModel/ContactValidator.cs]: contactInfo should probably be renamed to something like "text"

File: ContactViewModel/ContactViewModel/ContactViewModel.cs [hunk 4]
```diff
@@ -1,9 +1,26 @@
-﻿namespace ContactManagement.Views
+﻿using System.Collections.Generic;
```
Comment (c-j-j) [inline ContactViewModel/ContactViewModel/ContactViewModel.cs]: `ContactViewModel/ContactViewModel/ContactViewModel.cs` sounds a little odd as a namespace.

File: WPFGui/GuiContactManagement/AddContact/AddContactInputView.cs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        {
+            SetContactFields();
```
Comment (c-j-j) [inline WPFGui/GuiContactManagement/AddContact/AddContactInputView.cs]: was there a need to extract a method here?

File: WPFGui/GuiContactManagement/AddContact/AddContactInputView.cs [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        private void SetContactFields()
+        {
+            if (IsValidContactDetails(contactFieldPanel))
+            {
+                viewModel.SetFirstName(contactFieldPanel.FirstNameField());
+                viewModel.SetLastName(contactFieldPanel.LastNameField());
+                viewModel.SetStreetAddress(contactFieldPanel.StreetAddressField());
+                viewModel.SetPostcode(contactFieldPanel.PostcodeField());
+                viewModel.SetTelephoneNumber(contactFieldPanel.TelephoneNoField());
+                viewModel.SubmitContactToViewModel();
```
Comment (c-j-j) [inline WPFGui/GuiContactManagement/AddContact/AddContactInputView.cs]: is there a need to do all of this setting of fields? Why not just pass the ContactFieldPanel to SubmitContactToViewModel and have that extract the fields so we don't do this indirection?

File: WPFGui/GuiContactManagement/AddContact/AddContactViewModel.cs [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+{
+    public class AddContactViewModel
+    {
+        private readonly ContactViewModel.ContactViewModel viewModel;
+        private readonly Dictionary<string, string> contactInfo = new Dictionary<string, string>();
+
+        public AddContactViewModel(ContactViewModel.ContactViewModel viewModel)
+        {
+            this.viewModel = viewModel;
+        }
```
Comment (c-j-j) [inline WPFGui/GuiContactManagement/AddContact/AddContactViewModel.cs]: I'm not really seeing why this AddContactViewModel is necessary. Why not just use the  `ContactViewModel` directly as your adapter from GUI -> Core?

File: ContactViewModel/ContactViewModel/ContactViewModelTest.cs [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+        public void Setup()
+        {
+            fakeStorage = new FakeContactStorage();
+            contactViewModel = new ContactViewModel(fakeStorage);
+        }
+
+        [Test]
+        public void AddANewContactToStorage()
+        {
+            var fakeStorage = new FakeContactStorage();
```
Comment (jsuchy) [inline ContactViewModel/ContactViewModel/ContactViewModelTest.cs]: why do this initialisation again? maybe this was done first but forgot to remove the duplication? attention to these details is extremely important.

File: ContactViewModel/ContactViewModel/ContactViewModelTest.cs [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+            var contact1 = fakeStorage.ReadAllContacts().First();
+            var contact2 = fakeStorage.ReadAllContacts().Last();
+
+            AssertIsSameContact(foundContact1, contact1);
+            AssertIsSameContact(foundContact2, contact2);
+        }
+
+        private static void AssertIsSameContact(Contact firstContact, Contact secondContact)
+        {
+            Assert.That(firstContact.FirstName == secondContact.FirstName);
```
Comment (jsuchy) [inline ContactViewModel/ContactViewModel/ContactViewModelTest.cs]: would it make more sense for the Contact to be able to determine this?

File: ContactViewModel/ContactViewModel/ContactViewModelTest.cs [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+            var foundContact2 = contactViewModel.GetContacts().Last();
+
+            var contact1 = fakeStorage.ReadAllContacts().First();
+            var contact2 = fakeStorage.ReadAllContacts().Last();
+
+            AssertIsSameContact(foundContact1, contact1);
+            AssertIsSameContact(foundContact2, contact2);
+        }
+
+        private static void AssertIsSameContact(Contact firstContact, Contact secondContact)
```
Comment (jsuchy) [inline ContactViewModel/ContactViewModel/ContactViewModelTest.cs]: Why make these helper methods static?

File: ContactViewModel/ContactViewModel/ContactViewModelTest.cs [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+            const string postcode = "postcode";
+            const string telephone = "telephone";
+
+            contactViewModel.AddContact(firstName, lastName, street, postcode, telephone);
+
+            return new Contact(firstName, lastName, street, postcode, telephone);
+        }
+    }
+
+    internal class FakeContactStorage : IContactStorage
```
Comment (jsuchy) [inline ContactViewModel/ContactViewModel/ContactViewModelTest.cs]: why internal and not private? if this class is used elsewhere, it deserves to be a first-class citizen and have its own file.

**Pull request merged**